### PR TITLE
Adding documentation about scaledown transaction recovery processing

### PR DESCRIPTION
**Pull request by ochaloup**

### CODE REVIEW

File: doc/user-guide.adoc [hunk 1]
```diff
@@ -215,3 +215,101 @@ spec:
   disableHTTPRoute: true
 ----
 
+[[scaledown-transaction-recovery]]
+## Transaction recovery during scaledown
+
+As the application deployed in the WildFly application server
+may use JTA transactions there and question emerges: what happen when the cluster size is scaled down.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "what does happen when the cluster is scaled down?"

File: doc/user-guide.adoc [hunk 2]
```diff
@@ -215,3 +215,101 @@ spec:
...
+On decreasing of number of active WildFly replicas there may be still some in-doubt transactions running.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "When the number of active WildFly replicas is decreased, there may..."

File: doc/user-guide.adoc [hunk 3]
```diff
@@ -215,3 +215,101 @@ spec:
...
+When the pod is removed then all the in-progress transactions are stopped and rolled-back.
+A more troublesome situation occurs when XA transactions are used.
+When the XA transaction declares it's prepared it's a promise to finish the transaction successfully.
+But the transaction manager which made this promise is a running process inside of the WildFly server.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "... this promise is running inside the WildFly server"

File: doc/user-guide.adoc [hunk 4]
```diff
@@ -215,3 +215,101 @@ spec:
...
+Then just simply shutting down such pod would lead to a data inconsistencies or to fact of leaving locks on data.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "... load to data inconsistencies or data locks"

File: doc/user-guide.adoc [hunk 5]
```diff
@@ -215,3 +215,101 @@ spec:
...
+
+That's why it's needed to be ensured that all transactions are finished before the number of replicas is really decreased.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "It must be ensured that all ..."

File: doc/user-guide.adoc [hunk 6]
```diff
@@ -215,3 +215,101 @@ spec:
...
+For that purpose the WildFly operator provides scale down functionality which verifies if all transactions were finished
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "WildFly Operator"

File: doc/user-guide.adoc [hunk 7]
```diff
@@ -215,3 +215,101 @@ spec:
...
+
+```
+oc patch wildflyserver <<name>> -p '[{"op":"replace", "path":"/spec/replicas", "value":0}]' --type json
+```
+
+or you can manually edit and change the replica number with with `oc edit wildflyserver <<name>>`.
+
+NOTE: Decreasing replica size at the `StatefulSet` or deleting the Pod itself has no effect and as such changes will be reverted.
+
+WARNING: if you decide to delete whole `WildflyServer` definition (`oc delete wildflyserver <<deployment_name>>`)
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "`oc delete wildflyserver <deployment_name>`" (using << >> creates a relative link in asciidoc)

File: doc/user-guide.adoc [hunk 8]
```diff
@@ -215,3 +215,101 @@ spec:
...
+         then no transaction recovery process is started and the pod is terminated regardless of unfinished transactions.
+         If you want to remove the deployment and you expects doing it in transaction safe way
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "If you want to remove the deployment in a safe way without data inconsistencies, you need..."

File: doc/user-guide.adoc [hunk 9]
```diff
@@ -215,3 +215,101 @@ spec:
...
+         you need first to scale down the number of pods to 0,
+         wait until all pods are terminated
+         and only after that you can delete the `WildFlyServer` instance
+
+WARNING: Narayana recovery listener has to be enabled in Wildfly transaction subsystem.
+         Otherwise scaledown transaction recovery processing is skipped for the particular WildFly pod.
+         See the `recovery-listener` at
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "See the [ `recovery-listener` attribute of the transaction subsystem](https://wildscribe.github.io/WildFly/18.0/subsystem/transactions/index.html#attr-recovery-listener)."

File: doc/user-guide.adoc [hunk 10]
```diff
@@ -215,3 +215,101 @@ spec:
...
+         https://wildscribe.github.io/WildFly/18.0/subsystem/transactions/index.html.
+         The recovery listener is **not** enabled by default!
```
Comment (jmesnil) [inline doc/user-guide.adoc]: I thought it was enabled by default in the WildFly 18 S2I image?

File: doc/user-guide.adoc [hunk 11]
```diff
@@ -215,3 +215,101 @@ spec:
...
+
+when the scaledown process begins the pod state (`oc get pod <<pod_name>>``) is still marked as `Running`.
+The reason is that that the pod needs to be able to finish all the unfinished transactions and which includes the remote EJB calls that targets it.
+If you want to observe the state of the scaledown processing you need to observe the status of the `WildFlyServer` instance.
+When running `oc describe wildflyserver<<name>>` you can see the status of the Pods.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "`oc describe wildflyserver <name>`"

File: doc/user-guide.adoc [hunk 12]
```diff
@@ -215,3 +215,101 @@ spec:
...
+
+The `WildFlyServer.Status.Pods[].State` can be on of these
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "one of the following values:"

File: doc/user-guide.adoc [hunk 13]
```diff
@@ -215,3 +215,101 @@ spec:
...
+| The WildFly contain some unfinished transactions. The pod cannot be terminated until they are cleaned.
+  The transaction recovery is periodically run at WildFly and it waits the transactions are finished eventually.
+
+| SCALING_DOWN_CLEAN
+| The pod was processed by transaction scaled down processing and is marked as clean to be removed from the cluster.
+
+|===
+
+
+You can observe the overall state of the active and no-active pods when take a look at the
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "by looking at the `WildFlyServer.Status.ScalingdownPods` and `WildFlyServer.Status.Replicas` fields"

File: doc/user-guide.adoc [hunk 14]
```diff
@@ -215,3 +215,101 @@ spec:
...
+`WildFlyServer.Status.'Scalingdown Pods'` and `WildFlyServer.Status.Replicas`.
+First defines the number of pods which are about to be terminated when they are clean of unfinished transactions.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "`ScalingdownPods` defines ..."

File: doc/user-guide.adoc [hunk 15]
```diff
@@ -215,3 +215,101 @@ spec:
...
+The second one defines the current number of running pods.
```
Comment (jmesnil) [inline doc/user-guide.adoc]: "`Replicas` defines ..."

**Comment by jmesnil** [comment]
@ochaloup thanks, it looks good, I proposed some minor changes to clarify the doc.

My only actual question is https://github.com/wildfly/wildfly-operator/pull/122#discussion_r335392441

File: doc/user-guide.adoc (hunk 10 above)
Comment (ochaloup) [reply doc/user-guide.adoc]: Ok, that's right. For some reason I haven't recognized that. I was thinking in the scope of the WildFly default `standalone.xm`. I will remove the sentence.
Thanks.

File: doc/user-guide.adoc (hunk 7 above)
Comment (ochaloup) [reply doc/user-guide.adoc]: I see, I was thinking about it but I thought that when used in the code block `\`\`` it's taken in a different way. I will change that.

**Comment by ochaloup** [comment]
@jmesnil thanks a lot for the review. I updated the text.

File: doc/user-guide.adoc [hunk 16]
```diff
@@ -215,3 +215,98 @@ spec:
...
+If the transaction is in state of heuristics the pod is marked as `SCALING_DOWN_RECOVERY_DIRTY`
+and the administrator needs to manually connect with the `jboss-cli` to the particular WildFly instance
+and to resolve the heuristic transaction.
+
+When all the formerly heuristics records are removed from the transaction object store then the operator
+marks the pod as `SCALING_DOWN_CLEAN` and the pod is terminated.
+
+#### SCALING_DOWN_CLEAN state and StatefulSet behaviour
+
+There is a special case coming from the design of the `StatefulSet` that enusres that the network hostname is stable
```
Comment (mchoma) [inline doc/user-guide.adoc]: ensures

Comment (ochaloup) [reply doc/user-guide.adoc]: thanks!

**Comment by jmesnil** [comment]
@ochaloup thanks

**Pull request merged**